jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ConcurrentPool.java

```diff
@@ -382,16 +440,24 @@ public static class ConcurrentEntry<E> implements Entry<E>
         //    1+ -> multiplex count
         private final AtomicBiInteger state = new AtomicBiInteger(0, -1);
         private final ConcurrentPool<E> pool;
-        // The pooled object. This is not volatile as it is set once and then never changed.
-        // Other threads accessing must check the state field above first, so a good before/after
-        // relationship exists to make a memory barrier.
-        private E pooled;
+        // A hard ref to the pooled object; this is volatile as it is nulled out on every
+        // getPooled() call and re-set on release.
+        private volatile E released;
+        // A weak ref to the pooled object; the hard ref is nulled out when it is read
+        // and re-set when it is released, so if the pooled object is never released and gets
+        // garbage collected, this can be detected as this weak ref will be cleared.
+        private WeakReference<E> pooled;
 
         public ConcurrentEntry(ConcurrentPool<E> pool)
         {
             this.pool = pool;
         }
 
+        boolean isGarbage()
+        {
+            return pooled != null && pooled.get() == null;
```

Comment:
Not enough.

But even `released == null && pooled != null && pooled.get() == null` does not work, because just after `pooled.get() != null` the GC runs, nulls it out, and then the algorithm in `acquire()` would return an acquired entry with pooled == null.